packages/docusaurus-plugin-client-redirects/src/__tests__/redirectCreators.test.ts

```diff
@@ -0,0 +1,87 @@
+
+const createExtensionValidationTests = (
+  redirectCreatorFactory: (extensions: string[]) => RedirectsCreator,
+) => {
+  test('should reject empty extensions', () => {
+    expect(() => {
+      redirectCreatorFactory(['.html']);
+    }).toThrowErrorMatchingInlineSnapshot(
+      `"Extension=['.html'] contains a . (dot) and is not allowed. If the redirect extension system is not good enough for your usecase, you can create redirects yourself with the 'createRedirects' plugin option."`,
+    );
+  });
+  test('should reject extensions with .', () => {
+    expect(() => {
+      redirectCreatorFactory(['.html']);
+    }).toThrowErrorMatchingInlineSnapshot(
+      `"Extension=['.html'] contains a . (dot) and is not allowed. If the redirect extension system is not good enough for your usecase, you can create redirects yourself with the 'createRedirects' plugin option."`,
+    );
+  });
+  test('should reject extensions with /', () => {
+    expect(() => {
+      redirectCreatorFactory(['ht/ml']);
+    }).toThrowErrorMatchingInlineSnapshot(
+      `"Extension=['ht/ml'] contains a / and is not allowed. If the redirect extension system is not good enough for your usecase, you can create redirects yourself with the 'createRedirects' plugin option."`,
+    );
+  });
+  test('should reject extensions with illegal url char', () => {
+    expect(() => {
+      redirectCreatorFactory([',']);
+    }).toThrowErrorMatchingInlineSnapshot(
+      `"Extension=[','] contains invalid uri characters. If the redirect extension system is not good enough for your usecase, you can create redirects yourself with the 'createRedirects' plugin option."`,
+    );
+  });
+};
+
+describe('toExtensionsRedirectCreator', () => {
+  createExtensionValidationTests(toExtensionsRedirectCreator);
+
+  test('should create redirects from html/htm extensions', () => {
+    const redirectCreator = toExtensionsRedirectCreator(['html', 'htm']);
+    expect(redirectCreator('')).toEqual([]);
+    expect(redirectCreator('/')).toEqual([]);
+    expect(redirectCreator('/abc.html')).toEqual(['/abc']);
+    expect(redirectCreator('/abc.htm')).toEqual(['/abc']);
+    expect(redirectCreator('/abc.xyz')).toEqual([]);
+  });
+
+  test('should not create redirection for an empty extension array', () => {
+    const redirectCreator = toExtensionsRedirectCreator([]);
+    expect(redirectCreator('')).toEqual([]);
+    expect(redirectCreator('/')).toEqual([]);
+    expect(redirectCreator('/abc.html')).toEqual([]);
+  });
+});
+
+describe('fromExtensionsRedirectCreator', () => {
+  createExtensionValidationTests(fromExtensionsRedirectCreator);
+
+  test('should create redirects to html/htm extensions', () => {
+    const redirectCreator = fromExtensionsRedirectCreator(['html', 'htm']);
+    expect(redirectCreator('')).toEqual([]);
+    expect(redirectCreator('/')).toEqual([]);
+    expect(redirectCreator('/abc')).toEqual(['/abc.html', '/abc.htm']);
+    expect(redirectCreator('/def.html')).toEqual([]);
```

Comment:
@yangshun @lex111 I'm not totally sure of this behavior:

If user wants to redirect from html + htm extensions, and the route already ends with one of the extensions he wants to redirect from (like `/def.html` here), I guess It's not worth creating redirections from path `/def.html.html` and `/def.html.htm`, that would be weird.

It's quite unlikely to happen and probably doesn't matter but if you have an opinion I'm curious